Compiler/preproc/preprocessor.cpp

```diff
@@ -0,0 +1,429 @@
+            }
+            else
+            {
+                _macros.add(macroName.GetCStr(), line.GetCStr());
+            }
+        }
+        else if (directive == "undef")
+        {
+            String macroName = GetNextWord(line);
+            if (macroName.GetLength() == 0)
+            {
+                LogError(ErrorCode::MacroNameMissing);
+            }
+            else if (!_macros.contains(macroName.GetCStr()))
+            {
+                LogError(ErrorCode::MacroDoesNotExist, String::FromFormat("Macro '%s' is not defined", macroName.GetCStr()));
+            }
+            else
+            {
+                _macros.remove_name(macroName.GetCStr());
+            }
+        }
+        else if (directive == "error")
+        {
+            LogError(ErrorCode::UserDefinedError, String::FromFormat("User error: %s", line.GetCStr()));
+        }
+        else if ((directive == "sectionstart") || (directive == "sectionend"))
+        {
+            // do nothing -- 2.72 put these as markers in the script
+        }
+        else if ((directive == "region") || (directive == "endregion"))
+        {
+            // do nothing -- scintilla can fold it, so it can be used to organize the code
+        }
+        else
+        {
+            LogError(ErrorCode::UnknownPreprocessorDirective, String::FromFormat("Unknown preprocessor directive '%s'", directive.GetCStr()));
+        }
+
+        return String("");  // replace the directive with a blank line
+    }
+
+    void Preprocessor::DefineMacro(const String& name, const String& value)
+    {
+        _macros.add(name.GetCStr(), value.GetCStr());
+    }
+
+    String Preprocessor::PreProcessLine(const String& lineToProcess)
+    {
+        if (DeletingCurrentLine())
+        {
+            return String("");
+        }
+
+        std::vector<StringBuilder> previousOutput = std::vector<StringBuilder>();
+        std::vector<String> previousLine = std::vector<String>();
+        StringBuilder output = StringBuilder(lineToProcess.GetLength());
+        String line = lineToProcess;
+        std::vector<String> ignored = std::vector<String>();
+        while (line.GetLength() > 0)
+        {
+            size_t i = 0;
+            while ((i < line.GetLength()) && (!is_alphanum(line[i])))
+            {
+                if ((line[i] == '"') || (line[i] == '\''))
+                {
+                    i = line.FindChar(line[i],i);
+                    if (i == NOT_FOUND)
+                    {
+                        i = line.GetLength();
+                        break;
+                    }
+                }
+                i++;
+            }
+
+            output.WriteString(line.Mid(0, i));
+
+            if (i < line.GetLength())
+            {
+                bool precededByDot = false;
+                if (i > 0) precededByDot = (line[i - 1] == '.');
+
+                line = line.Mid(i, line.GetLength() - i);
```

Comment:
Not much important, just noticed this randomly, and wanted to note that you can write simply `line.Mid(i)` here. When the `count` argument is not passed the Mid will take everything to the end.

Hmm, also the whole `line = line.Mid(i);` may be done as `line.ClipLeft(i);`